src/modules/sensors/vehicle_magnetometer/VehicleMagnetometer.cpp

```diff
@@ -256,88 +260,111 @@ void VehicleMagnetometer::Run()
 				sub.unregisterCallback();
 			}
 
-			if (_selected_sensor_sub_index >= 0) {
-				PX4_INFO("%s switch from #%u -> #%d", "MAG", _selected_sensor_sub_index, best_index);
+			if (_param_sens_mag_mode.get() == 1) {
+				if (_selected_sensor_sub_index >= 0) {
+					PX4_INFO("%s switch from #%u -> #%d", "MAG", _selected_sensor_sub_index, best_index);
+				}
 			}
 
 			_selected_sensor_sub_index = best_index;
 			_sensor_sub[_selected_sensor_sub_index].registerCallback();
 		}
 	}
 
-	if ((_selected_sensor_sub_index >= 0)
-	    && (_voter.get_sensor_state(_selected_sensor_sub_index) == DataValidator::ERROR_FLAG_NO_ERROR)
-	    && updated[_selected_sensor_sub_index]) {
-
-		const sensor_mag_s &mag = _last_data[_selected_sensor_sub_index];
-
-		_mag_timestamp_sum += mag.timestamp_sample;
-		_mag_sum += Vector3f{mag.x, mag.y, mag.z};
-		_mag_sum_count++;
+	// Publish
+	if (_param_sens_mag_mode.get() == 1) {
+		// publish only best mag
+		if ((_selected_sensor_sub_index >= 0)
+		    && (_voter.get_sensor_state(_selected_sensor_sub_index) == DataValidator::ERROR_FLAG_NO_ERROR)
+		    && updated[_selected_sensor_sub_index]) {
 
-		if ((_param_sens_mag_rate.get() > 0)
-		    && hrt_elapsed_time(&_last_publication_timestamp) >= (1e6f / _param_sens_mag_rate.get())) {
-
-			const Vector3f magnetometer_data = _mag_sum / _mag_sum_count;
-			const hrt_abstime timestamp_sample = _mag_timestamp_sum / _mag_sum_count;
-
-			// reset
-			_mag_timestamp_sum = 0;
-			_mag_sum.zero();
-			_mag_sum_count = 0;
-
-			// populate vehicle_magnetometer with primary mag and publish
-			vehicle_magnetometer_s out{};
-			out.timestamp_sample = timestamp_sample;
-			out.device_id = mag.device_id;
-			magnetometer_data.copyTo(out.magnetometer_ga);
-
-			out.timestamp = hrt_absolute_time();
-			_vehicle_magnetometer_pub.publish(out);
+			Publish(_selected_sensor_sub_index);
+		}
 
-			_last_publication_timestamp = out.timestamp;
+	} else if (_param_sens_mag_mode.get() == 0) {
+		for (int uorb_index = 0; uorb_index < MAX_SENSOR_COUNT; uorb_index++) {
+			// publish all magnetometers as separate instances
+			if (updated[uorb_index] && (_calibration[uorb_index].device_id() != 0)) {
+				Publish(uorb_index, true);
```

Comment:
Do we care about the advertisement odering?

Reply:
I don't think so, but I suppose we could try to keep it consistent for simplicity with listener or log review. I've been going out of my way to always use the device ids and never depend on order. It's really only

Reply:
I've updated it to advertise up to and including the corresponding vehicle_magnetometer publication instance as sensor_mags are discovered/added.